### ISSUE TITLE
can this work in Linux?

### ISSUE THREAD

**Issue by heckeBBG** [MolecularMatters/raw_pdb#9]
We are thinking of using this under Linux to parse PDB files. Has this library been compiled and run under Linux by anyone?

**Comment by MolecularMatters** [comment]
Someone on reddit gave it a try and needed a few changes (though I don't know which), but I don't think it would be a lot of work.

You'd have to replace the _BitScanForward intrinsic and maybe change some minor things to make GCC happy, but it shouldn't be much.

The library does not pull in any Windows-specific header.

Let us know how it works out, a PR would be appreciated.

Am 11. März 2022 19:15:47 MEZ schrieb heckeBBG ***@***.***>:
>We are thinking of using this under Linux to parse PDB files. Has this library been compiled and run under Linux by anyone?
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/MolecularMatters/raw_pdb/issues/9
>You are receiving this because you are subscribed to this thread.
>
>Message ID: ***@***.***>
-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by tromey** [comment]
> Someone on reddit gave it a try and needed a few changes (though I don't know which)
That was me.

I tried again and needed a few tweaks.  Mostly I hacked around things, so I doubt you'd want my patch verbatim.  Also, it would be simpler perhaps to use clang, some of the changes are just to handle GCC.

> You'd have to replace the _BitScanForward intrinsic
You can replace it with `ffsl` on Linux.

> The library does not pull in any Windows-specific header.
Actually I think `<intrin.h>` is Windows-specific.  I guess it's related to the bit scan thing.

Also there's this:
```
	// this matches the definition in guiddef.h, but we don't want to pull that in
	struct GUID
	{
		unsigned long  Data1;
		unsigned short Data2;
		unsigned short Data3;
		unsigned char  Data4[8];
	};

	static_assert(sizeof(GUID) == 16u, "Size mismatch.");
```
This causes an assert on Linux:
```
PDB_Types.h:26:36: error: static assertion failed: Size mismatch.
   26 |         static_assert(sizeof(GUID) == 16u, "Size mismatch.");
      |                       ~~~~~~~~~~~~~^~~~~~
```
There's another assert as well:

**Comment by tromey** [comment]
Dang it, hit enter too soon.
```
PDB_Types.h:47:52: error: static assertion failed: Size mismatch.
   47 |         static_assert(sizeof(IMAGE_SECTION_HEADER) == 40u, "Size mismatch.");
      |                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~
```
In particular the `GUID` type comes out as:
```
(gdb) ptype/o GUID
/* offset      |    size */  type = struct GUID {
/*      0      |       8 */    unsigned long Data1;
/*      8      |       2 */    unsigned short Data2;
/*     10      |       2 */    unsigned short Data3;
/*     12      |       8 */    unsigned char Data4[8];
/* XXX  4-byte padding   */

                               /* total size (bytes):   24 */
                             }
```

**Comment by tromey** [comment]
I suppose using `uint32_t` instead of `unsigned long` and `uint16_t` instead of `unsigned short` makes it closer (on Linux) to what's intended.  But I wonder if all the types in PDB_Types.h are layout-sensitive.

**Comment by MolecularMatters** [comment]
> I suppose using `uint32_t` instead of `unsigned long` and `uint16_t` instead of `unsigned short` makes it closer (on Linux) to what's intended. But I wonder if all the types in PDB_Types.h are layout-sensitive.

Yes, that definitely should be changed, both for the IMAGE_SECTION_HEADER and GUID declarations. Those were taken verbatim from the corresponding headers, but should be replaced with declarations using types from cstdint.

Other type declarations should be fine, since they're all defined in terms of cstdint types.
I can add static_asserts though to make sure.

**Comment by tromey** [comment]
Worth noting that if this is mapping directly from the file, then for a full port to Linux it will have to byte-swap the data for big-endian hosts.  With newer versions of gcc you can use the `scalar_storage_order` attribute to do this without extra source, though there are some limitations.

**Comment by heckeBBG** [comment]
Thank you everybody for the pointers. We got it to compile and a simple test program executed. Should we end up pursuing this seriously (and thus, verify it is actually solid), we will provide a PR
